### PR TITLE
add admin server repocheck

### PR DESCRIPTION
**Pull request by MaximilianMeister**
reuse repocheck command to not only check the addons but to check
crowbar as well.
rename addons to components to fit the new context with crowbar.

depends on:
https://github.com/crowbar/crowbar-core/pull/630

### CODE REVIEW

**Comment by MaximilianMeister** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;otxWJSSF&#x2F;96-3-s47p10-implement-api-for-repo-check-for-admin-server-upgrade-step-4"> (3) S47P10 Implement API for repo check for admin server (upgrade step 4)</a>

**Comment by jsuchome** [comment]
I'm not sure if we should mix those cases into one command. 
- Admin server repocheck uses zypper to check products present on the admin server. 
- AddOns check is to check if Addons are prepared for use by nodes. For this, I think we should use the repo library (like we did in 5-6 upgrade)

@vuntz ?

**Comment by MaximilianMeister** [comment]
> I'm not sure if we should mix those cases into one command.

we don't mix anything, we are just calling a different API endpoint

> Admin server repocheck uses zypper to check products present on the admin server.
> AddOns check is to check if Addons are prepared for use by nodes. For this, I think we should use the repo library (like we did in 5-6 upgrade)

it doesn't matter as this happens on the server side, the client doesn't need to know about it

**Comment by rsalevsky** [comment]
+1

**Comment by MaximilianMeister** [comment]
@jsuchome @Itxaka anyone time for a quick review?

File: lib/crowbar/client/app/upgrade.rb [hunk 1]
```diff
@@ -219,12 +219,12 @@ def backup(component)
           catch_errors(e)
         end
 
-        desc "repocheck ADDON",
-          "Check for existing repositories for an addon (ha|storage)"
+        desc "repocheck COMPONENT",
+          "Check for existing repositories for a component (crowbar|ha|storage)"
```
Comment (jsuchome) [inline lib/crowbar/client/app/upgrade.rb]: _crowbar_ is a bit confusing here. I understand it calls a method from `crowbar` module, but to user, this should be shown as a check for Cloud product repository

Comment (MaximilianMeister) [inline lib/crowbar/client/app/upgrade.rb]: to me it is not confusing at all. so not only cloud, but also sles12sp2 is checked, so that are the repos that crowbar itself needs, so why not call the component `crowbar`?

Comment (MaximilianMeister) [inline lib/crowbar/client/app/upgrade.rb]: > I understand it calls a method from crowbar module

this herre is completely independent of the model, I have not even thought about it

Comment (vuntz) [inline lib/crowbar/client/app/upgrade.rb]: I actually believe we're checking for openstack more than crowbar ;-)

Comment (jsuchome) [inline lib/crowbar/client/app/upgrade.rb]: > so why not call the component crowbar?

We do not have anything like crowbar repository.

Comment (MaximilianMeister) [inline lib/crowbar/client/app/upgrade.rb]: > I actually believe we're checking for openstack more than crowbar ;-)

somehow yes, but what about `sles` ?

Comment (MaximilianMeister) [inline lib/crowbar/client/app/upgrade.rb]: > We do not have anything like crowbar repository.

in fact we're upgrading the admin (crowbar) node, that's why i think it makes sense to call it `crowbar`.

Comment (vuntz) [inline lib/crowbar/client/app/upgrade.rb]: If it's for the repocheck for the admin server, it's fine. If it's the repocheck for the node upgrade, then I'd go with openstack/ha/ceph.

Comment (MaximilianMeister) [inline lib/crowbar/client/app/upgrade.rb]: @jsuchome are you fine with the name too?

Comment (jsuchome) [inline lib/crowbar/client/app/upgrade.rb]: Not really :-( 
Could you explain in `long_desc`, what's `crowbar` value supposed to mean, so that potential user does not have to search the code?

Comment (MaximilianMeister) [inline lib/crowbar/client/app/upgrade.rb]: > Not really :-(

then how would you name it?

> Could you explain in long_desc, what's crowbar value supposed to mean, so that potential user does not have to search the code?

good idea, I will add some more description

**Comment by jsuchome** [comment]
Thanks, +1

File: lib/crowbar/client/app/upgrade.rb [hunk 2]
```diff
@@ -219,12 +219,21 @@ def backup(component)
...
+          for a specific component. COMPONENT can be 'crowbar', 'storage' or 'ha'
+
+          crowbar: checks for the repositories required to upgrade the crowbar
+          server
+
+          storage: checks for the repositories required to upgrade the nodes that
+          have the SUSE Enterprise Storage Addon installed
+
+          ha: checks for the repositories required to upgrade the nodes that
+          have the SUSE Enterprise High Availability Addon installed
```
Comment (vuntz) [inline lib/crowbar/client/app/upgrade.rb]: So what should be used to check the repo required to upgrade nodes that don't need HA/Ceph?

Comment (MaximilianMeister) [inline lib/crowbar/client/app/upgrade.rb]: good question, you mean only checking for the `os` feature?

Comment (MaximilianMeister) [inline lib/crowbar/client/app/upgrade.rb]: @vuntz I guess we need a basic check, which, as you said, could be called `openstack`. but for that we still don't have an API. would you be fine to add it in another PR?

Comment (vuntz) [inline lib/crowbar/client/app/upgrade.rb]: I'm fine with it in another PR, but see comment I'll add after this one.

Btw, SUSE Enterprise High Availability doesn't exist. It's SLE HA, not SE HA.

**Comment by vuntz** [comment]
What I find very confusing here is that we put at the same level the repocheck for the admin server and the repocheck for the products needed for nodes, while what the repocheck APIs return is not necessarily the same thing.

(it's actually debatable whether the two sets of API should have a similar name, which is the cause of the confusion)

**Comment by rsalevsky** [comment]
I think the repo check for the admin node is fine like it is. But we shouldn't check explicit for HA or Storage we should just check for the node repos and if HA or Storage is used we should return that they are missing.

Maybe like this:
```
crowbarctl repocheck nodes

Repo       Status
OS          ok
HA           ok
Storage   failed
```

**Comment by MaximilianMeister** [comment]
> crowbarctl repocheck nodes

sounds good to me what @rsalevsky suggested. lets continue in a followup PR

File: lib/crowbar/client/app/upgrade.rb (hunk 2 above)
Comment (MaximilianMeister) [inline lib/crowbar/client/app/upgrade.rb]: fixed, thanks

**Comment by MaximilianMeister** [comment]
the follow up PR for the backend is https://github.com/crowbar/crowbar-core/pull/644

once this is merged I can adapt `crowbarctl`, by removing the 'ha' and 'ceph' options, and replace it with 'nodes'. the 'crowbar' repocheck here is ok, so I'll merge for now
